contracts/swaps/connectors/SwapsImplSovrynSwap.sol

```diff
@@ -0,0 +1,176 @@
+                );
+            }
+        }
+
+    }
+
+    /**
+     * check is the existing allowance suffices to transfer the needed amount of tokens.
+     * if not, allows the transfer of an arbitrary amount of tokens.
+     * @param tokenAmount the amount to transfer
+     * @param tokenAddress the address of the token to transfer
+     * @param sovrynSwapNetwork the address of the sovrynSwap network contract.
+     * */
+    function allowTransfer(uint256 tokenAmount, address tokenAddress, address sovrynSwapNetwork) internal{
+        uint256 tempAllowance = IERC20(tokenAddress).allowance(address(this), sovrynSwapNetwork);
+        if (tempAllowance < tokenAmount) {
+            IERC20(tokenAddress).safeApprove(
+                sovrynSwapNetwork,
+                uint256(-1)
+            );
+        }
+    }
+
+    /**
+     * calculates the number of source tokens to provide in order to obtain the required destination amount.
+     * @param sourceTokenAddress the address of the source token address
+     * @param destTokenAddress the address of the destination token address
+     * @param requiredDestTokenAmount the number of destination tokens needed
+     * @param maxSourceTokenAmount the maximum number of source tokens to spend
+     * @return the estimated amount of source tokens needed. minimum: minSourceTokenAmount, maximum: maxSourceTokenAmount
+     * */
+    function estimateSourceTokenAmount(address sourceTokenAddress, address destTokenAddress, uint requiredDestTokenAmount,  uint maxSourceTokenAmount) internal returns(uint256 estimatedSourceAmount){
+
+        uint256 sourceToDestPrecision = IPriceFeeds(priceFeeds).queryPrecision(sourceTokenAddress, destTokenAddress);
+        if (sourceToDestPrecision == 0)
+            return maxSourceTokenAmount;
+
+        //compute the expected rate for the maxSourceTokenAmount -> if spending less, we can't get a worse rate.
+        uint256 expectedRate = internalExpectedRate(sourceTokenAddress, destTokenAddress, maxSourceTokenAmount);
+
+        //compute the source tokens needed to get the required amount with the worst case rate
+        estimatedSourceAmount = requiredDestTokenAmount
+            .mul(sourceToDestPrecision)
+            .div(expectedRate);
+
+
+        //never spend more than the maximum
+        if (estimatedSourceAmount == 0 || estimatedSourceAmount > maxSourceTokenAmount)
+            return maxSourceTokenAmount;
+
+    }
+
+    /**
+     * returns the expected rate for 1 source token when exchanging the given amount of source tokens
+     * @param sourceTokenAddress the address of the source token contract
+     * @param destTokenAddress the address of the destination token contract
+     * @param sourceTokenAmount the amount of source tokens to get the rate for
+     * */
+    function internalExpectedRate(
+        address sourceTokenAddress,
+        address destTokenAddress,
+        uint256 sourceTokenAmount)
+        public
```

Comment:
is this ok? The function name is `internal`

Reply:
same as with the internalSwap.